src/test/spec/unified_runner/matcher.rs

```diff
@@ -0,0 +1,342 @@
+    actual: Option<&Bson>,
+    expected: &Bson,
+    returns_root_documents: bool,
+    entities: Option<&EntityMap>,
+) -> bool {
+    results_match_inner(actual, expected, returns_root_documents, true, entities)
+}
+
+pub fn results_match_inner(
+    actual: Option<&Bson>,
+    expected: &Bson,
+    returns_root_documents: bool,
+    root: bool,
+    entities: Option<&EntityMap>,
+) -> bool {
+    match expected {
+        Bson::Document(expected_doc) => {
+            if let Some((key, value)) = expected_doc.iter().next() {
+                if key.starts_with("$$") && expected_doc.len() == 1 {
+                    return special_operator_matches((key, value), actual, entities);
+                }
+            }
+
+            let actual_doc = match actual {
+                Some(Bson::Document(actual)) => actual,
+                // The only case in which None is an acceptable value is if the expected document
+                // is a special operator; otherwise, the two documents do not match.
+                _ => return false,
+            };
+
+            for (key, value) in expected_doc {
+                if !results_match_inner(actual_doc.get(key), value, false, false, entities) {
+                    return false;
+                }
+            }
+
+            // Documents that are not the root-level document should not contain extra keys.
+            if !root {
+                for (key, _) in actual_doc {
+                    if !expected_doc.contains_key(key) {
+                        return false;
+                    }
+                }
+            }
+
+            true
+        }
+        Bson::Array(expected_array) => {
+            let actual_array = actual.unwrap().as_array().unwrap();
+            if expected_array.len() != actual_array.len() {
+                return false;
+            }
+
+            // Some operations return an array of documents that should be treated as root
+            // documents.
+            for (actual, expected) in actual_array.iter().zip(expected_array) {
+                if !results_match_inner(
+                    Some(actual),
+                    expected,
+                    false,
+                    returns_root_documents,
+                    entities,
+                ) {
+                    return false;
+                }
+            }
+
+            true
+        }
+        Bson::Int32(_) | Bson::Int64(_) | Bson::Double(_) => match actual {
+            Some(actual) => numbers_match(actual, expected),
+            None => false,
+        },
+        _ => match actual {
+            Some(actual) => actual == expected,
+            None => false,
+        },
+    }
+}
+
+fn numbers_match(actual: &Bson, expected: &Bson) -> bool {
+    if actual.element_type() == expected.element_type() {
+        return actual == expected;
+    }
+
+    match (actual, expected) {
+        (Bson::Int32(actual), Bson::Int64(expected)) => i32_matches_i64(*actual, *expected),
```

Comment:
because each of these pairs always includes an integer, we can use `bson_util::get_int` to reduce some of the boilerplate here.